src/WinRT.Runtime/ComWrappersSupport.net5.cs

```diff
@@ -99,12 +106,15 @@ public static T CreateRcwForComObject<T>(IntPtr ptr)
                 winrtObj.Resurrect();
             }
 
-            return rcw switch
-            {
-                ABI.System.Nullable<string> ns => (T)(object) ns.Value,
-                ABI.System.Nullable<Type> nt => (T)(object) nt.Value,
-                _ => (T) rcw
-            };
+            return rcw switch
+            {
+                ABI.System.Nullable<string> ns => (T)(object)ns.Value,
+                ABI.System.Nullable<Type> nt => (T)(object)nt.Value,
+                T castRcw => castRcw,
+                _ when tryUseCache => CreateRcwForComObject<T>(ptr, false),
+                _ => throw new InvalidCastException()

```

Comment:
I wonder if the last fallback should be `(T) rcw` to allow us to get a proper exception message on what is being casted from what and to what.

Reply:
It might also be worth using a different exception type with a custom message so the failure case doesn't look like a bug in C#/WinRT.